plugins/inputs/docker_log/docker_logs.go

```diff
@@ -0,0 +1,440 @@
+	d.containerList[containerId] = logReader
+	return nil
+}
+
+func (d *DockerLogs) removeFromContainerList(containerId string) error {
+	delete(d.containerList, containerId)
+	return nil
+}
+
+func (d *DockerLogs) containerInContainerList(containerId string) bool {
+	if _, ok := d.containerList[containerId]; ok {
+		return true
+	}
+	return false
+}
+
+func (d *DockerLogs) stopAllReaders() error {
+	for _, container := range d.containerList {
+		container.Close()
+	}
+	return nil
+}
+
+func (d *DockerLogs) containerListUpdate(acc telegraf.Accumulator) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout.Duration)
+	defer cancel()
+	if d.client == nil {
+		log.Printf("%s : Dock client is null", ERR_PREFIX)
+		return nil
+	}
+	containers, err := d.client.ContainerList(ctx, d.opts)
+	if err != nil {
+		log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+		return err
+	}
+	for _, container := range containers {
+		if d.containerInContainerList(container.ID) {
+			continue
+		}
+		d.wg.Add(1)
+		/*Start a new goroutine for every new container that has logs to collect*/
+		go func(c types.Container) {
+			defer d.wg.Done()
+			err := d.getContainerLogs(c, acc)
+			if err != nil {
+				log.Printf("%s : %s ", ERR_PREFIX, err.Error())
+			}
+			d.removeFromContainerList(c.ID)
+			return
+		}(container)
+	}
+	return nil
+}
+
+func (d *DockerLogs) getContainerLogs(
+	container types.Container,
+	acc telegraf.Accumulator,
+) error {
+	c, err := d.client.ContainerInspect(context.Background(), container.ID)
+	if err != nil {
+		log.Printf("%s : %s", ERR_PREFIX, err.Error())
+		return err
+	}
+	logOptions := types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Timestamps: false,
+		Details:    true,
+		Follow:     true,
+		Tail:       "0",
+	}
+	logReader, err := d.client.ContainerLogs(context.Background(), container.ID, logOptions)
+	d.addToContainerList(container.ID, logReader)
+	if err != nil {
+		log.Printf("%s : %s", ERR_PREFIX, err.Error())
+		return err
+	}
+	/* Parse container name */
+	cname := "unknown"
+	if len(container.Names) > 0 {
+		/*Pick first container name*/
+		cname = strings.TrimPrefix(container.Names[0], "/")
+	}
+
+	tags := map[string]string{
+		"containerId":   container.ID,
+		"containerName": cname,
+	}
+	// Add labels to tags
+	for k, label := range container.Labels {
+		if d.labelFilter.Match(k) {
+			tags[k] = label
+		}
+	}
+	if c.Config.Tty {
+		_, err = pushTtyLogs(acc, tags, logReader)
+	} else {
+		_, err = pushLogs(acc, tags, logReader)
+	}
+	if err != nil {
+		log.Printf("%s : %s", ERR_PREFIX, err.Error())
+		return err
+	}
+	return nil
+}
+func pushTtyLogs(acc telegraf.Accumulator, tags map[string]string, src io.Reader) (written int64, err error) {
+	tags["logType"] = "unknown" //in tty mode we wont be able to differentiate b/w stdout and stderr hence unknown
+	fields := map[string]interface{}{}
+	data := make([]byte, LOG_BYTES_MAX)
+	for {
+		num, err := src.Read(data)
```

Comment:
Read returns a stream which may or may not be one of more log "lines" . Therefore it is read as a stream until EOF. So its not techincally a line. Its just a stream..

Reply:
I think we need to wrap it with bufio.Scanner then.

Reply:
In https://github.com/moby/moby/blob/master/pkg/stdcopy/stdcopy.go , src is an io.Reader and they seem to be reading it the way it is currently(as well as handling EOF) , they aren't using bufio.Scanner . Also bufio.Scanner seems to be using newline termination as the default token , would that be a good idea if there are dockers that just spit out logs to /dev/stdout without newline termination?

Reply:
Without the Scanner or similar splitting, the message emitted could be anything: just the start of a line, then end of a line, the middle, or multiple lines.  After all, it is just a stream as is clear from reading [stdcopy.go#L152](https://github.com/moby/moby/blob/master/pkg/stdcopy/stdcopy.go#L152).

Like you mention, not doing line splitting does allow for non-line delimited data, and splitting on lines could use lots of memory if the line length is not limited, but I think this is the exception and that virtually all use will be line oriented, byte streaming is probably not what is wanted here.

So this makes me think, if we want this plugin to read log lines then it acts very much like the tail plugin, only with a different source.  This makes me think we should pass the data through a parser like we do in tail.  I think this would be worth adding and its not much work, what do you think?  It would be possible to do something very similar to the current implementation with the value parser, but it would also make it much easier to use this plugin with the grok parser.

Reply:
I still feel its best to rely on taking what the io.Reader buffer has at a particular point and pushing that out as one entry. I was thinking of other cases , like for example : since this code deals with docker console output. if someone does an "ls -lt" command , each file or folder entry of the ls command would be a new line and would that cause an over burst of entries ?If there are 1000 files in a directory and I run this command , would it give 1000 entries?
 
 Having said that , If you still insist on the Scanner option , we could try and approach it. I went through the value and grok parsers . What generic grok pattern do you plan to use for the scanner output? I'll have to spend a bit of time to understand how to integrate these parsers into this . If you feel you would be able to integrate them much faster since you already have a good understanding on the available parsers , feel free to make necessary changes!

Reply:
If we added support for parsers it would be up to the plugin user to provide a custom grok pattern, it would need to match for all containers they are tailing so it would probably be paired with a restrictive container name pattern.  It's not too much work, but that said, I talked this over with @glinton, and we came to the conclusion that we should probably hold off on adding parser support, and for now a user can always use the parser processor separately.

I do think we need the Scanner though, when it comes to logging it always seems desirable to collect as full lines and not potentially only partial bits of a line.  I think most of the time there will be no difference, lines will be flushed one at a time, but right now you could imagine some pathological cases like one metric per character.  Ultimately, we have to segment the messages and I think one per line is better than undefined.  I'll give implementing this a shot tomorrow and see how it goes.